### ISSUE TITLE
Integration of User- & Item-Features

### ISSUE THREAD

**Issue by MatthiasKirsch** [lyst/lightfm#221]
Hi, 

I have a problem with integrating user- and item-features.
After some research I now know that I have to concatenate my user/item-feature matrix with an identity matrix. After doing this, my user-matrix looks the following: 
![grafik](https://user-images.githubusercontent.com/18574614/31308232-5b8794ee-ab73-11e7-9915-2e1d66c5a20e.png)

When I know integrate this matrix in csr-format into my model fit it with (user_features = my_features) export the top-N recommendation using the predict function (with user_features agian!) I get the following precision when evaluating it. 
![figure_1](https://user-images.githubusercontent.com/18574614/31308252-a9938a1c-ab73-11e7-9e09-f4116322a28b.png)
The blue line here is without adding User_features!

What is going wrong there? :( 

My code looks like this:

```
NUM_COMPONENTS = 10
NUM_EPOCHS = 20
learn_rate = 0.02
NUM_THREADS = 10

model = LightFM(loss = 'bpr, no_components=NUM_COMPONENTS, learning_rate=learn_rate)

model.fit(train, epochs=NUM_EPOCHS, num_threads=10, user_features=user_features)
```
`
after doing this I export the top-N for every User in my testset:

```
recommendations = list()
n_users, n_items = test.shape
for user_id in np.arrange(n_users):
     scores = model.predict(user_id, np.arange(n_items), num_threads=10, user_features=user_features)
     top_items_list = np.argsort(-scores)[:20].tolist()
     productids = itemgetter(*top_items_list)(products)
     userid = customers[user_id]
     recommendations.append({'UserID': userid, 'Empfehlungen': productids, 'Scores': np.sort(scores)[::-1 [:20].tolist()})
```

Here customers are the real labels of the customers.

**Comment by maciejkula** [comment]
It looks like the hyperparameters which are appropriate for the model without user features aren't appropriate for the model with user features.

1. Are your user features all binary? If not, do they have large (in absolute value) values? If so, you need to scale them to lie between 0 and 1, or in a similar small range.
2. Have you tried a lower learning rate, or performed any hyperparameter searches?

**Comment by MatthiasKirsch** [comment]
1.	My user features are all binary yes. They look very similar to the image I posted above. So I did one hot encoding for “my own user features” and concatenated these matrices via hstack. So for example if a user_1 is male and young, he has a 1 in the columns gender_male and age_young and of cause in the identity matrix for himself.
2.	I tried it tonight with these features: 
NUM_THREADS = 10
NUM_COMPONENTS = 30 (before: 10)
NUM_EPOCHS = 30 (before: 20)
ITEM_ALPHA = 1e-8 (do I need this when using item_features?)
USER_ALPHA = ITEM_ALPHA (do I need this when using user_features?)
So I left the learning rate of in this scenario. I also added item_features into the process and the result looks like this: (the blue line is again without side information) 
![grafik](https://user-images.githubusercontent.com/18574614/31315044-690c03ba-ac0f-11e7-8c72-24dc4eab5de9.png)

And I did not do any hyperparameter searches yet. I only saw the one here http://lyst.github.io/lightfm/docs/examples/learning_schedules.html?highlight=hyperparameter which explains how to handle the EPOCHS I think. And in #189 you commented “To select the right values, I would define an evaluation function and run a hyperparameter search, for example using sklearn's ParameterSampler.“, but I don’t really know yet how to do that. 
3.	Ok I will also try to test it with reduced number of THREADS in my parameters.
Some extra information: My data is some sparse purchase data (5000000 transactions) and I have about 500000 users with attributes gender and age. There is no missing data in my user_features.

**Comment by maciejkula** [comment]
(Aside: you can use triple backticks (```) to format code blocks.)

1. Hyperparameter search is one of the most important concepts to master in machine learning. There are lots of good examples and guides on the internet (like the  [sklearn one](http://scikit-learn.org/stable/modules/grid_search.html)). I highly recommend you look at those. For LightFM, the principle will be the same: you will run your model on the same training data with different hyperparameters, and look at the accuracy; then select that set of hyperparameters that gives best accuracy.
2. I suggested the learning rate specifically because I think it has the best chance of helping you. Please try it.
3. Are you sure that your matrix concatenation works? Not doing the concatenation correctly would produce results very similar to the ones you have.

The side information you have is fairly sparse, so I wouldn't really expect it to help, but the model should be no worse than with no side information. Tweaking the learning rate is I think key.

**Comment by MatthiasKirsch** [comment]
Ok thanks!I will have a look on how to perform hyperparameter search for LightFM to get the learning_rate that gives best accuray.

This is how I produced the user_features matrix: (data is my complete dataframe and customers is a list of unique sorted UserIDs  `customers = list(np.sort(data.UserID.unique()))`

```
def create_user_features(customers, data):
    data = data[["UserID", "Gender", "Age_Group"]].drop_duplicates().reset_index(drop=True)
    sorterIndex = dict(zip(customers, range(len(customers)))) 
    data['User_Rank'] = data['UserID'].map(sorterIndex) 
    data = data.sort_values(by='User_Rank')
    data = data[['Gender', 'Age_Group']]
    data_as_dicts = [dict(r.iteritems()) for _, r in data.iterrows()]
    vectorizer = DictVectorizer()
    user_feat = vectorizer.fit_transform(data_as_dicts)
    user_iden = identity(len(customers), format='csr')
    user_features = sparse.hstack([user_iden, user_feat])
    return user_features
```

**Comment by maciejkula** [comment]
This looks fine, as far as I can tell. Try lowering the learning rate!

**Comment by maciejkula** [comment]
Without having the data I may not be able to help more.

One possibility is that you do not actually stand to gain from your user features: in actual fact, they seem very coarse and maybe not that helpful. In LightFM, user latent representations are unweighted sums of the latent representations of their features: here, you are probably simply averaging something informative (a specific user's latent feature) with very uninformative features (gender and age), and the resulting model is worse.

**Comment by MatthiasKirsch** [comment]
Ok thanks! Perhaps it could be usefull not to group the ages in "young", "middle" and "old" but their normal ages like 18, 20, 30... or perhaps to group them in a more wide spectrum (e. g. 10, 20, 30, 40, ...)?

Would you try it?

**Comment by maciejkula** [comment]
It's possible that it might help, but even so I think the features would not carry much information.

**Comment by bhangaboy** [comment]
@MatthiasKirsch I modified your function to generate item features. When i use those item features in model.fit
i get this error
**Incorrect number of features in item_features**

**Comment by KJenia** [comment]
hi!
I couldn't understand one thing, after model.fit could I make model.predict for users_id which was'n in model fit? 
for example, I fit the model for user_id = [1,2,3] 
and I want make prediction for user_id=4,  is it possible do it without retrain the model?
p.s. of course there is no new user or item features for predicted user
thanks for suggestion
